internal/sftp/goclient_test.go

```diff
@@ -0,0 +1,252 @@
+package sftp_test
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gliderlabs/ssh"
+	gosftp "github.com/pkg/sftp"
+	gossh "golang.org/x/crypto/ssh"
+	"gotest.tools/v3/assert"
+	tfs "gotest.tools/v3/fs"
+
+	"github.com/artefactual-sdps/enduro/internal/sftp"
+)
+
+// PubkeyHandler handles checking the client's public key against the keys in
+// the authorized_keys file.
+func pubkeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
+	file, err := os.Open("./testdata/authorized_keys")
+	if err != nil {
+		log.Fatalln("SSH: couldn't open authorized_keys file.")
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		allowed, _, _, _, err := ssh.ParseAuthorizedKey([]byte(scanner.Text()))
+		if err != nil {
+			log.Fatalln("SSH: couldn't parse authorized key.")
+		}
+		if ssh.KeysEqual(key, allowed) {
+			return true
+		}
+	}
+
+	log.Println("SSH: unknown key provided.")
+	return false
+}
+
+// HostKeySigner signs messages from the server to the client and allows the
+// client to confirm the host key signature.
+func hostKeySigner() (gossh.Signer, error) {
+	keyfile := "./testdata/serverkeys/test_rsa"
+
+	key, err := os.ReadFile(keyfile)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't read keyfile %q, %v\n", keyfile, err)
+	}
+
+	signer, err := gossh.ParsePrivateKey(key)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't parse private key: %v\n", err)
+	}
+
+	return signer, nil
+}
+
+// SftpHandler starts the SFTP subsystem.
+func sftpHandler(sess ssh.Session) {
+	debugStream := io.Discard
+	serverOptions := []gosftp.ServerOption{
+		gosftp.WithDebug(debugStream),
+	}
+	server, err := gosftp.NewServer(
+		sess,
+		serverOptions...,
+	)
+	if err != nil {
+		log.Printf("sftp server init error: %s\n", err)
+		return
+	}
+	if err := server.Serve(); err == io.EOF {
+		server.Close()
+		fmt.Println("sftp client exited session.")
+	} else if err != nil {
+		fmt.Println("sftp server completed with error:", err)
+	}
+}
+
+// StartSFTPServer starts a test SFTP server, and returns a pointer to the
+// server. The caller must call Close() to shut down the server when done with
+// it.
+func startSFTPServer() (*ssh.Server, error) {
+	srv := ssh.Server{
+		Addr: "127.0.0.1:2222",
+		Handler: func(s ssh.Session) {
+			authorizedKey := gossh.MarshalAuthorizedKey(s.PublicKey())
+			io.WriteString(s, fmt.Sprintf("public key used by %s:\n", s.User()))
+			s.Write(authorizedKey)
+		},
+		PublicKeyHandler: pubkeyHandler,
+		SubsystemHandlers: map[string]ssh.SubsystemHandler{
+			"sftp": sftpHandler,
+		},
+	}
+
+	signer, err := hostKeySigner()
+	if err != nil {
+		return nil, err
+	}
+	srv.AddHostKey(signer)
+
+	go func() {
+		err = srv.ListenAndServe()
+	}()
+
+	return &srv, err
```

Comment:
Very nice! Thanks! :heart_eyes:

Reply:
Turns out you can't call `t.Fatal(err)` from a goroutine: `call to (*T).Fatal from a non-test goroutine`

Reply:
Fair enough! I'd try with an error channel instead if it doesn't take you a lot of time. It could be useful to see the error message, e.g. say the developer already another process listening on `:2222`, it'd be hard to tell that's the problem if you don't capture the error.

I haven't tested, but something like this may work:

```go
errCh := make(chan error, 1)
go func() {
    // Non-blocking send because the channel is buffered.
    errCh <- srv.ListenAndServe()
}()

for {
    select {
    case err := <-errCh:
        if err != nil {
            t.Fatal(err)
        }
        return
    default:
        conn, err := net.DialTimeout("tcp", "127.0.0.1:2222", time.Second)
        if err == nil {
            conn.Close()
            return
        }
        time.Sleep(10 * time.Millisecond)
    }
}
```